salt_stack/salt/lambda-dev/files/build_lambda.py

```diff
@@ -0,0 +1,303 @@
+    """
+    print("----------------------------------------------------------------------------------")
+    print(cmd)
+
+    env = os.environ.copy()
+    env['STAGING_DIR'] = staging_dir
+
+    proc = subprocess.Popen(['/bin/bash'],
+                            env=env,
+                            cwd=staging_dir,
+                            bufsize=1, # line bufferred
+                            stderr=subprocess.STDOUT,
+                            stdout=subprocess.PIPE,
+                            stdin=subprocess.PIPE)
+
+    proc.stdin.write(cmd.encode())
+    proc.stdin.close()
+
+    for line in proc.stdout:
+        print(line.decode('utf8'), end='', flush=True)
+
+    # Verify that the command has finished cleaning up
+    # sometime stdout/stderr are closed but the command
+    # hasn't finished cleanup by the first poll() call
+    while proc.poll() is None:
+        time.sleep(1)
+
+    if proc.poll() != 0:
+        raise Exception("Return code: {}".format(proc.returncode))
+
+def run(cmd):
+    """Run the given command
+
+    Note: This method will stream the results of the command to stdout
+          so that for long running commands the results can be seen
+          before the command finishes
+
+    Args:
+        cmd (str): The command to run
+
+    Raises:
+        Exception: If the return code is non-zero
+    """
+    # DP TODO: With a little massaging script and run could be merged into a single method
+    print("----------------------------------------------------------------------------------")
+    print(cmd)
+
+    proc = subprocess.Popen(shlex.split(cmd),
+                            bufsize=1, # line bufferred
+                            stderr=subprocess.STDOUT,
+                            stdout=subprocess.PIPE)
+
+    for line in proc.stdout:
+        print(line.decode('utf8'), end='', flush=True)
+
+
+    # Verify that the command has finished cleaning up
+    # sometime stdout/stderr are closed but the command
+    # hasn't finished cleanup by the first poll() call
+    while proc.poll() is None:
+        time.sleep(1)
+
+    if proc.poll() != 0:
+        raise Exception("Return code: {}".format(proc.returncode))
+
+if __name__ == '__main__':
+    cur_dir = pathlib.Path(__file__).parent
+    os.chdir(cur_dir)
+
+    if len(sys.argv) != 3:
+        print("Usage: {} <domain name> <bucket name>".format(sys.argv[0]))
+        sys.exit(-1)
+
+    domain = sys.argv[1]
+    bucket = sys.argv[2]
+
+    # Not all AWS Lambda containers have these libraries installed
+    # verify that they are installed before importing them
+    # (This can happen with non-python runtimes)
+    # DP NOTE: using --user in case the script is not run as root
+    #          this can sometimes result in an import error, though
+    #          running the script again (after the packages have been
+    #          installed) normally works
+    run('python3 -m pip install --user boto3 PyYaml')
```

Comment:
You're creating  a `boto3` user elsewhere (I'm just scrolling through the files in order, so might not have come across that yet).

Reply:
A boto3 user? This was added because when running the NodeJS container it has python installed but not boto3 or PyYaml. This allows this build script run in the NodeJS container.

Reply:
I was reading boto3 as the parameter of `--user`. I claim tiredness. 😁